1. Install Dependencies

In [1]:
!pip install -U transformers datasets peft trl bitsandbytes accelerate evaluate rouge-score nltk absl-py

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.9/72.9 MB 49.4 MB/s eta 0:00:00:00:0100:01
  Attempting uninstall: bitsandbytes
    Found existing installation: bitsandbytes 0.46.0
    Uninstalling bitsandbytes-0.46.0:
      Successfully uninstalled bitsandbytes-0.46.0


 2. Hugging Face Login

In [17]:

from huggingface_hub import login
login(token="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")

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 714c50453200, raw_cell="
from huggingface_hub import login
login(token="hf.." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for post_run_cell), with arguments args (<ExecutionResult object at 714c50460b60, execution_count=17 error_before_exec=None error_in_exec=None info=<ExecutionInfo object at 714c50453200, raw_cell="
from huggingface_hub import login
login(token="hf.." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

Load Dataset

In [3]:
from datasets import load_dataset

dataset = load_dataset("azizshaw/text_to_json")

In [4]:
dataset

DatasetDict({
    train: Dataset({
        features: ['input', 'output', 'instruction'],
        num_rows: 845
    })
})

In [5]:
def format_example(example):
    return {
        "text": f"Convert the following text into JSON:\n\n{example['instruction']}\n\nJSON:\n{example['output']}"
    }

train_data = dataset["train"].map(format_example)

4. Load LLaMA-3.1-8B-Instruct + Tokenizer

In [8]:
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from peft import prepare_model_for_kbit_training, LoraConfig, get_peft_model, TaskType
import torch

model_name = "meta-llama/Meta-Llama-3.1-8B-Instruct"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    torch_dtype=torch.bfloat16,
    quantization_config=bnb_config
)

model = prepare_model_for_kbit_training(model)


Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

Loading checkpoint shards: 100%|██████████| 4/4 [00:02<00:00,  1.34it/s]


In [9]:
model = prepare_model_for_kbit_training(model)

lora_config = LoraConfig(
    r=8,
    lora_alpha=16,
    target_modules=["q_proj", "v_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type=TaskType.CAUSAL_LM,
)

model = get_peft_model(model, lora_config)


5. Tokenizer

In [10]:
tokenizer = AutoTokenizer.from_pretrained(model_name)
tokenizer.pad_token = tokenizer.eos_token


6. SFT Training

In [14]:
from trl import SFTTrainer, SFTConfig

sft_config = SFTConfig(
    output_dir="./llama3-8b-json",
    per_device_train_batch_size=8,
    gradient_checkpointing=True,
    bf16=True,
    logging_steps=10,
    num_train_epochs=4,
    save_strategy="epoch"
)

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 714c5041a030, raw_cell="from trl import SFTTrainer, SFTConfig

sft_config .." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#X20sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for post_run_cell), with arguments args (<ExecutionResult object at 714af81e3d70, execution_count=14 error_before_exec=None error_in_exec=None info=<ExecutionInfo object at 714c5041a030, raw_cell="from trl import SFTTrainer, SFTConfig

sft_config .." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#X20sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

In [15]:
def formatting_func(example):
    return example["text"]

trainer = SFTTrainer(
    model=model,
    
    args=sft_config,
    train_dataset=train_data,
    formatting_func=formatting_func,
    
)

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 714c504510d0, raw_cell="def formatting_func(example):
    return example[".." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#X21sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

No label_names provided for model class `PeftModelForCausalLM`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for post_run_cell), with arguments args (<ExecutionResult object at 714c50452e40, execution_count=15 error_before_exec=None error_in_exec=None info=<ExecutionInfo object at 714c504510d0, raw_cell="def formatting_func(example):
    return example[".." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#X21sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

In [16]:
trainer.train()

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 714d18c4e720, raw_cell="trainer.train()" store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#X22sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

BrokenPipeError: [Errno 32] Broken pipe

Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x714af8075bb0>> (for post_run_cell), with arguments args (<ExecutionResult object at 714c50446300, execution_count=16 error_before_exec=None error_in_exec=[Errno 32] Broken pipe info=<ExecutionInfo object at 714d18c4e720, raw_cell="trainer.train()" store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#X22sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

Evaluation

In [16]:
from evaluate import load
from tqdm import tqdm
import json

rouge = load("rouge")

finetuned_model = trainer.model

In [17]:
def evaluate_model(model, tokenizer, data, check_json=False):
    predictions = []
    references = []
    json_valid_count = 0

    model.eval()

    for example in tqdm(data, desc="Evaluating"):
        prompt = f"Convert the following text into JSON:\n\n{example['instruction']}\n\nJSON:"
        input_ids = tokenizer(prompt, return_tensors="pt", truncation=True).input_ids.to(model.device)

        with torch.no_grad():
            output_ids = model.generate(
                input_ids,
                max_new_tokens=256,
                do_sample=False,
                temperature=0.0
            )

        output = tokenizer.decode(output_ids[0], skip_special_tokens=True)
        generated_json = output.split("JSON:")[-1].strip()

        predictions.append(generated_json)
        references.append(example["output"])

        if check_json:
            try:
                json.loads(generated_json)
                json_valid_count += 1
            except json.JSONDecodeError:
                pass

    scores = rouge.compute(predictions=predictions, references=references)
    if check_json:
        scores["valid_json_percent"] = round(100 * json_valid_count / len(predictions), 2)

    return scores

Load Base Model for Comparison

In [18]:
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    torch_dtype=torch.bfloat16,
    quantization_config=bnb_config
)

Loading checkpoint shards: 100%|██████████| 4/4 [00:03<00:00,  1.30it/s]


In [19]:
base_tokenizer = AutoTokenizer.from_pretrained(model_name)
base_tokenizer.pad_token = base_tokenizer.eos_token


Run Evaluation

In [20]:
sample_eval_data = dataset["test"] if "test" in dataset else dataset["train"].select(range(10))

results_base = evaluate_model(base_model, base_tokenizer, sample_eval_data)
results_finetuned = evaluate_model(finetuned_model, tokenizer, sample_eval_data, check_json=True)

print("\n🔹 Base Model:", results_base)
print("✅ Fine-tuned Model:", results_finetuned)

Evaluating:   0%|          | 0/10 [00:00<?, ?it/s]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Evaluating:  10%|█         | 1/10 [00:16<02:24, 16.11s/it]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your


🔹 Base Model: {'rouge1': np.float64(0.7414812624732058), 'rouge2': np.float64(0.7382972276927737), 'rougeL': np.float64(0.7414812624732057), 'rougeLsum': np.float64(0.5105698869096016)}
✅ Fine-tuned Model: {'rouge1': np.float64(0.724916021615306), 'rouge2': np.float64(0.6599326599326599), 'rougeL': np.float64(0.7127208996640865), 'rougeLsum': np.float64(0.49539944501241423), 'valid_json_percent': 0.0}


Base Model: {'rouge1': np.float64(0.7414812624732058), 'rouge2': np.float64(0.7382972276927737), 'rougeL': np.float64(0.7414812624732057), 'rougeLsum': np.float64(0.5105698869096016)}

✅ Fine-tuned Model: {'rouge1': np.float64(0.724916021615306), 'rouge2': np.float64(0.6599326599326599), 'rougeL': np.float64(0.7127208996640865), 'rougeLsum': np.float64(0.49539944501241423), 'valid_json_percent': 0.0}

SFT training (for 3 epoch and 8 batch size)

In [1]:
from trl import SFTTrainer, SFTConfig

sft_config = SFTConfig(
    output_dir="./llama3-8b-json",
    per_device_train_batch_size=8,
    gradient_checkpointing=True,
    bf16=True,
    logging_steps=10,
    num_train_epochs=3,
    save_strategy="epoch"
)

/home/ubuntu/anaconda3/envs/genai/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [10]:
def formatting_func(example):
    return example["text"]

trainer = SFTTrainer(
    model=model,
    
    args=sft_config,
    train_dataset=train_data,
    formatting_func=formatting_func,
    
)

No label_names provided for model class `PeftModelForCausalLM`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


In [11]:
trainer.train()

wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Currently logged in as: prajaktapatil328 (prajaktapatil328-blue-polaris) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.


wandb: WARNING The get_url method is deprecated and will be removed in a future release. Please use `run.url` instead.


TrainOutput(global_step=318, training_loss=0.6026377630121303, metrics={'train_runtime': 3106.4434, 'train_samples_per_second': 0.816, 'train_steps_per_second': 0.102, 'total_flos': 1.1694257970610176e+17, 'train_loss': 0.6026377630121303})

In [12]:
finetuned_model = trainer.model

Evalution

In [13]:
from evaluate import load
from tqdm import tqdm
import json

rouge = load("rouge")

finetuned_model = trainer.model

In [14]:
def evaluate_model(model, tokenizer, data, check_json=False):
    predictions = []
    references = []
    json_valid_count = 0

    model.eval()

    for example in tqdm(data, desc="Evaluating"):
        prompt = f"Convert the following text into JSON:\n\n{example['instruction']}\n\nJSON:"
        input_ids = tokenizer(prompt, return_tensors="pt", truncation=True).input_ids.to(model.device)

        with torch.no_grad():
            output_ids = model.generate(
                input_ids,
                max_new_tokens=256,
                do_sample=False,
                temperature=0.0
            )

        output = tokenizer.decode(output_ids[0], skip_special_tokens=True)
        generated_json = output.split("JSON:")[-1].strip()

        predictions.append(generated_json)
        references.append(example["output"])

        if check_json:
            try:
                json.loads(generated_json)
                json_valid_count += 1
            except json.JSONDecodeError:
                pass

    scores = rouge.compute(predictions=predictions, references=references)
    if check_json:
        scores["valid_json_percent"] = round(100 * json_valid_count / len(predictions), 2)

    return scores

Load base  model for comparsiom

In [15]:
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    torch_dtype=torch.bfloat16,
    quantization_config=bnb_config
)

Loading checkpoint shards: 100%|██████████| 4/4 [00:13<00:00,  3.36s/it]


In [16]:
base_tokenizer = AutoTokenizer.from_pretrained(model_name)
base_tokenizer.pad_token = base_tokenizer.eos_token

Run Evalution

In [17]:
sample_eval_data = dataset["test"] if "test" in dataset else dataset["train"].select(range(10))

results_base = evaluate_model(base_model, base_tokenizer, sample_eval_data)
results_finetuned = evaluate_model(finetuned_model, tokenizer, sample_eval_data, check_json=True)

print("\n🔹 Base Model:", results_base)
print("✅ Fine-tuned Model:", results_finetuned)

Evaluating:   0%|          | 0/10 [00:00<?, ?it/s]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Evaluating:  10%|█         | 1/10 [00:13<02:02, 13.58s/it]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your


🔹 Base Model: {'rouge1': np.float64(0.7414812624732058), 'rouge2': np.float64(0.7382972276927737), 'rougeL': np.float64(0.7414812624732057), 'rougeLsum': np.float64(0.5105698869096016)}
✅ Fine-tuned Model: {'rouge1': np.float64(0.7415433403805496), 'rouge2': np.float64(0.7021291952363767), 'rougeL': np.float64(0.7354827343199435), 'rougeLsum': np.float64(0.4926708949964764), 'valid_json_percent': 0.0}


In [18]:
import pandas as pd
pd.DataFrame([results_base, results_finetuned], index=["Base", "Fine-tuned"])


rouge1    rouge2    rougeL  rougeLsum  valid_json_percent
Base        0.741481  0.738297  0.741481   0.510570                 NaN
Fine-tuned  0.741543  0.702129  0.735483   0.492671                 0.0

SFT training (for 5 epoch and 8 batch size)

In [30]:
from trl import SFTTrainer, SFTConfig

sft_config = SFTConfig(
    output_dir="./llama3-8b-json",
    per_device_train_batch_size=8,
    gradient_checkpointing=True,
    bf16=True,
    logging_steps=10,
    num_train_epochs=5,
    save_strategy="epoch"
)

In [45]:
def formatting_func(example):
    return example["text"]

trainer2 = SFTTrainer(
    model=model,
    
    args=sft_config,
    train_dataset=train_data,
    formatting_func=formatting_func,
    
)

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 742095ec5610, raw_cell="def formatting_func(example):
    return example[".." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y112sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

No label_names provided for model class `PeftModelForCausalLM`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for post_run_cell), with arguments args (<ExecutionResult object at 742095ec7410, execution_count=45 error_before_exec=None error_in_exec=None info=<ExecutionInfo object at 742095ec5610, raw_cell="def formatting_func(example):
    return example[".." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y112sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

In [44]:
trainer2.train()

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 7421f1b19400, raw_cell="trainer2.train()" store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y113sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

KeyboardInterrupt: 

Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for post_run_cell), with arguments args (<ExecutionResult object at 7421f1b1a900, execution_count=44 error_before_exec=None error_in_exec= info=<ExecutionInfo object at 7421f1b19400, raw_cell="trainer2.train()" store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y113sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

In [43]:
finetuned_model = trainer2.model

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 7421f1b18530, raw_cell="finetuned_model = trainer2.model" store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y114sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for post_run_cell), with arguments args (<ExecutionResult object at 7421f1b1bdd0, execution_count=43 error_before_exec=None error_in_exec=None info=<ExecutionInfo object at 7421f1b18530, raw_cell="finetuned_model = trainer2.model" store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y114sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

In [42]:
from evaluate import load
from tqdm import tqdm
import json

rouge = load("rouge")

finetuned_model = trainer2.model

Error in callback <bound method _WandbInit._pre_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for pre_run_cell), with arguments args (<ExecutionInfo object at 7421f1b14c20, raw_cell="from evaluate import load
from tqdm import tqdm
im.." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y116sdnNjb2RlLXJlbW90ZQ%3D%3D>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

Error in callback <bound method _WandbInit._post_run_cell_hook of <wandb.sdk.wandb_init._WandbInit object at 0x74243030bb00>> (for post_run_cell), with arguments args (<ExecutionResult object at 7421f1b15580, execution_count=42 error_before_exec=None error_in_exec=None info=<ExecutionInfo object at 7421f1b14c20, raw_cell="from evaluate import load
from tqdm import tqdm
im.." store_history=True silent=False shell_futures=True cell_id=vscode-notebook-cell://ssh-remote%2Bec2-34-235-207-189.compute-1.amazonaws.com/home/ubuntu/prajakta/llama.3.1-8b%28Instruct%29.ipynb#Y116sdnNjb2RlLXJlbW90ZQ%3D%3D> result=None>,),kwargs {}:


BrokenPipeError: [Errno 32] Broken pipe

SFT training (for 9 epoch and 8 batch size)

In [19]:
from trl import SFTTrainer, SFTConfig

sft_config = SFTConfig(
    output_dir="./llama3-8b-json",
    per_device_train_batch_size=8,
    gradient_checkpointing=True,
    bf16=True,
    logging_steps=10,
    num_train_epochs=9,
    save_strategy="epoch"
)

In [21]:
def formatting_func(example):
    return example["text"]

trainer1 = SFTTrainer(
    model=model,
    
    args=sft_config,
    train_dataset=train_data,
    formatting_func=formatting_func,
    
)

No label_names provided for model class `PeftModelForCausalLM`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


In [22]:
trainer1.train()

TrainOutput(global_step=954, training_loss=0.008635353703106474, metrics={'train_runtime': 9313.574, 'train_samples_per_second': 0.817, 'train_steps_per_second': 0.102, 'total_flos': 3.508277391183053e+17, 'train_loss': 0.008635353703106474})

In [23]:
finetuned_model = trainer1.model

Evalution

In [35]:
from evaluate import load
from tqdm import tqdm
import json

rouge = load("rouge")

finetuned_model = trainer1.model

In [36]:
def evaluate_model(model, tokenizer, data, check_json=False):
    predictions = []
    references = []
    json_valid_count = 0

    model.eval()

    for example in tqdm(data, desc="Evaluating"):
        prompt = f"Convert the following text into JSON:\n\n{example['instruction']}\n\nJSON:"
        input_ids = tokenizer(prompt, return_tensors="pt", truncation=True).input_ids.to(model.device)

        with torch.no_grad():
            output_ids = model.generate(
                input_ids,
                max_new_tokens=256,
                do_sample=False,
                temperature=0.0
            )

        output = tokenizer.decode(output_ids[0], skip_special_tokens=True)
        generated_json = output.split("JSON:")[-1].strip()

        predictions.append(generated_json)
        references.append(example["output"])

        if check_json:
            try:
                json.loads(generated_json)
                json_valid_count += 1
            except json.JSONDecodeError:
                pass

    scores = rouge.compute(predictions=predictions, references=references)
    if check_json:
        scores["valid_json_percent"] = round(100 * json_valid_count / len(predictions), 2)

    return scores

Loading base model for comaprsion

In [37]:
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    torch_dtype=torch.bfloat16,
    quantization_config=bnb_config
)

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

Loading checkpoint shards: 100%|██████████| 4/4 [00:13<00:00,  3.33s/it]


In [38]:
base_tokenizer = AutoTokenizer.from_pretrained(model_name)
base_tokenizer.pad_token = base_tokenizer.eos_token

Run Evalution

In [39]:
sample_eval_data = dataset["test"] if "test" in dataset else dataset["train"].select(range(10))

results_base = evaluate_model(base_model, base_tokenizer, sample_eval_data)
results_finetuned = evaluate_model(finetuned_model, tokenizer, sample_eval_data, check_json=True)

print("\n🔹 Base Model:", results_base)
print("✅ Fine-tuned Model:", results_finetuned)

Evaluating:   0%|          | 0/10 [00:00<?, ?it/s]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Evaluating:  10%|█         | 1/10 [00:13<02:04, 13.80s/it]The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
Evaluating:  20%|██        | 2/10 [00:27<01:49, 13.72s/it]The following generation flags are not valid


🔹 Base Model: {'rouge1': np.float64(0.7414812624732058), 'rouge2': np.float64(0.7382972276927737), 'rougeL': np.float64(0.7414812624732057), 'rougeLsum': np.float64(0.5105698869096016)}
✅ Fine-tuned Model: {'rouge1': np.float64(0.6907395358924313), 'rouge2': np.float64(0.6022441119834099), 'rougeL': np.float64(0.6505096508349599), 'rougeLsum': np.float64(0.38299718065495547), 'valid_json_percent': 0.0}


In [40]:
import pandas as pd
pd.DataFrame([results_base, results_finetuned], index=["Base", "Fine-tuned"])

rouge1    rouge2    rougeL  rougeLsum  valid_json_percent
Base        0.741481  0.738297  0.741481   0.510570                 NaN
Fine-tuned  0.690740  0.602244  0.650510   0.382997                 0.0